### PR TITLE
Add sf and hclust

### PR DESCRIPTION
**Pull request by mikemahoney218**
This PR does two things (though if it's easier, I can revert the hclust elements and push up a separate branch):

1. Adds sf to Imports and adds an sf method to `spatial_clustering_cv()`. This method ignores `coords` (with a warning if provided) in order to always calculate distance appropriately, no matter the CRS. This feels better conceptually, though I don't know if the clusters will always be "better":

``` r
library(spatialsample)
data(ames, package = "modeldata")
set.seed(123)
default <- spatial_clustering_cv(ames, coords = c(Latitude, Longitude), v = 5)

set.seed(123)
ames_sf <- sf::st_as_sf(ames, coords = c("Longitude", "Latitude"), crs = 4326)
new <- spatial_clustering_cv(ames_sf, v = 5)

library(ggplot2)
do.call(  
  rbind, 
  lapply(
    1:5, 
    \(x) cbind(assessment(default$splits[[x]]), fold = letters[[x]])
  )
) |> 
  ggplot(aes(Longitude, Latitude, color = fold)) + 
  geom_point() + 
  coord_sf(crs = 4326) + 
  theme_minimal() + 
  labs(title = "Non-sf clustering")
```

``` r
do.call(
  rbind, 
  lapply(
    1:5, 
    \(x) cbind(assessment(new$splits[[x]]), fold = letters[[x]])
  )
) |> 
    ggplot(aes(color = fold)) + 
    geom_sf() + 
    theme_minimal() + 
    labs(title = "sf clustering")
```

<sup>Created on 2022-05-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

As a side effect, it now passes a `dist` object to `kmeans` for clustering (rather than the coordinates themselves). I'd need a second opinion about if this makes sense. 

2. Adds a new argument (to both methods), `fun`, which allows users to specify which clustering method to use. Currently this is one of either `kmeans` or `hclust`; most results when searching "spatial clustering R" use hclust while acknowledging the clusters might not be inherently meaningful (eg https://gis.stackexchange.com/questions/17638/clustering-spatial-data-in-r ), though I haven't had much luck finding a paper/method citation for where this preference originated. 

``` r
library(spatialsample)
data(ames, package = "modeldata")
set.seed(123)
default <- spatial_clustering_cv(ames, 
                                 coords = c(Latitude, Longitude), 
                                 v = 5,
                                 fun = "hclust")

set.seed(123)
ames_sf <- sf::st_as_sf(ames, coords = c("Longitude", "Latitude"), crs = 4326)
new <- spatial_clustering_cv(ames_sf, v = 5, fun = "hclust")

library(ggplot2)
do.call(  
  rbind, 
  lapply(
    1:5, 
    \(x) cbind(assessment(default$splits[[x]]), fold = letters[[x]])
  )
) |> 
  ggplot(aes(Longitude, Latitude, color = fold)) + 
  geom_point() + 
  coord_sf(crs = 4326) + 
  theme_minimal() + 
  labs(title = "Non-sf clustering")
```

``` r
do.call(
  rbind, 
  lapply(
    1:5, 
    \(x) cbind(assessment(new$splits[[x]]), fold = letters[[x]])
  )
) |> 
    ggplot(aes(color = fold)) + 
    geom_sf() + 
    theme_minimal() + 
    labs(title = "sf clustering")
```

<sup>Created on 2022-05-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: R/spatial_clustering_cv.R [hunk 1]
```diff
@@ -32,17 +39,67 @@
...
-spatial_clustering_cv <- function(data, coords, v = 10, ...) {
-  coords <- tidyselect::eval_select(rlang::enquo(coords), data = data)
+spatial_clustering_cv <- function(data, coords, v = 10, fun = c("kmeans", "hclust"), ...) {
+  UseMethod("spatial_clustering_cv", data)
+}
 
+#' @rdname spatial_clustering_cv
+#' @export
+spatial_clustering_cv.default <- function(data, coords, v = 10, fun = c("kmeans", "hclust"), ...) {
+  fun <- match.arg(fun)
```
Comment (hfrick) [inline R/spatial_clustering_cv.R:60]: `rlang::arg_match()` is a bit stricter (no partial matching) and the error messages obey tidyverse standards so I would use that here

File: R/spatial_clustering_cv.R [hunk 2]
```diff
@@ -32,17 +39,67 @@
...
+                             v = v,
+                             fun = fun,
+                             ...)
+
+}
+
+#' @rdname spatial_clustering_cv
+#' @export
+spatial_clustering_cv.sf <- function(data, coords, v = 10, fun = c("kmeans", "hclust"), ...) {
+  fun <- match.arg(fun)
```
Comment (hfrick) [inline R/spatial_clustering_cv.R:80]: same comment about `rlang::arg_match()`

**Review by hfrick — CHANGES_REQUESTED** [review]
Looks good overall! A few high-level things:

- Passing the `dist` object to `kmeans()` instead of the coord data changes the resulting clustering. That doesn't necessarily mean that we can't do that but we should have a good reason for it. What was your motivation for changing that?

- The other rsample/spatialsample functions for cross-validation etc are not generics. To keep design patterns stable across the package, let's consider refactoring that aspect. 

- Generally, separating things like the support for `sf` and adding `hclust` into different PRs makes reviewing them easier. No need to change it now but for future reference :+1:

File: R/spatial_clustering_cv.R [hunk 3]
```diff
@@ -9,17 +9,24 @@
...
-#' @param data A data frame.
+#' @param data A data frame, or an `sf` object (often from [sf::read_sf()]
+#' or [sf::st_as_sf()]), to split into folds.
 #' @param coords A vector of variable names, typically spatial coordinates,
 #'  to partition the data into disjointed sets via k-means clustering.
+#'  This argument is ignored (with a warning) if `data` is an `sf` object.
 #' @param v The number of partitions of the data set.
-#' @param ... Extra arguments passed on to [stats::kmeans()].
-#' @export
+#' @param fun Which function to use to cluster. Must be one of either
```
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R:25]: `method` in particular conflicts with an argument in `hclust` which I think we want people to be able to pass `...` to. Is `type` too vague? `cluster_type`? `cluster_function`?

File: R/spatial_clustering_cv.R [hunk 4]
```diff
@@ -32,17 +39,67 @@
...
 #' @examples
-#' data(ames, package = "modeldata")
-#' spatial_clustering_cv(ames, coords = c(Latitude, Longitude), v = 5)
+#' data(Smithsonian, package = "modeldata")
+#' spatial_clustering_cv(Smithsonian, coords = c(latitude, longitude), v = 5)
+#'
+#' # When providing sf objects, coords are inferred automatically
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
+#'                         coords = c("longitude", "latitude"),
+#'                         crs = 4326) # WGS84
```
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R:48]: gah, having such trouble with auto-indentation in RStudio these days -- thank you

File: R/spatial_clustering_cv.R [hunk 5]
```diff
@@ -32,17 +39,67 @@
 #' pp. 5372-5375, doi: 10.1109/IGARSS.2012.6352393.
 #'
 #' @examples
-#' data(ames, package = "modeldata")
-#' spatial_clustering_cv(ames, coords = c(Latitude, Longitude), v = 5)
+#' data(Smithsonian, package = "modeldata")
+#' spatial_clustering_cv(Smithsonian, coords = c(latitude, longitude), v = 5)
+#'
+#' # When providing sf objects, coords are inferred automatically
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
```
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R:46]: I moved this lower to where we actually call `spatial_clustering_cv` -- the intent is to highlight "don't pass anything to the `coords` argument, we will not use it". Let me know if it's still out of place!

File: R/spatial_clustering_cv.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R:24]: cluster_function for now

**Comment by mikemahoney218** [comment]
@juliasilge I don't think GitHub will let me "re-request" you -- they probably want me to be more patient :laughing: -- but this should be good whenever you get time.

File: NEWS.md [hunk 6]
```diff
@@ -1,5 +1,19 @@
 # spatialsample (development version)
 
+* `spatial_clustering_cv()` gains an argument, `cluster_function`, which 
+  specifies what type of clustering to perform. `cluster_function = "kmeans"`, 
+  the default, uses [stats::kmeans()] for k-means clustering, while 
+  `cluster_function = "hclust"` uses [stats::hclust()] for hierarchical
```
Comment (juliasilge) [inline NEWS.md:6]:
```suggestion
  `cluster_function = "hclust"` uses `stats::hclust()` for hierarchical 
```
This won't link to anything in a NEWS file

File: R/spatial_clustering_cv.R [hunk 7]
```diff
@@ -9,17 +9,23 @@
...
-#'  the data into disjointed sets, as outlined in Brenning (2012). These
+#'  the data into disjointed sets, as outlined in Brenning (2012), or for
+#'  hierarchical clustering of the data. These
 #'  clusters are used as the folds for cross-validation. Depending on how the
 #'  data are distributed spatially, there may not be an equal number of points
 #'  in each fold.
 #'
-#' @param data A data frame.
+#' @param data A data frame, or an `sf` object (often from [sf::read_sf()]
+#' or [sf::st_as_sf()]), to split into folds.
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:19]:
```suggestion
#' @param data A data frame or an `sf` object (often from [sf::read_sf()]
#' or [sf::st_as_sf()]), to split into folds.
```

File: R/spatial_clustering_cv.R [hunk 8]
```diff
@@ -9,17 +9,23 @@
...
 #' @param coords A vector of variable names, typically spatial coordinates,
 #'  to partition the data into disjointed sets via k-means clustering.
+#'  This argument is ignored (with a warning) if `data` is an `sf` object.
 #' @param v The number of partitions of the data set.
-#' @param ... Extra arguments passed on to [stats::kmeans()].
-#' @export
+#' @param cluster_function Which function to use to cluster. Must be one of either
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:24]:
```suggestion
#' @param cluster_function Which function to use for clustering. Must be one of either
```

File: R/spatial_clustering_cv.R [hunk 9]
```diff
@@ -32,17 +38,42 @@
...
+#' spatial_clustering_cv(Smithsonian, coords = c(latitude, longitude), v = 5)
+#'
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
+#'                                coords = c("longitude", "latitude"),
+#'                                # Set CRS to WGS84
+#'                                crs = 4326)
+#'
+#' # When providing sf objects, coords are inferred automatically
+#' spatial_clustering_cv(smithsonian_sf, v = 5)
+#'
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:51]:
```suggestion
#' spatial_clustering_cv(smithsonian_sf, v = 5)
#' # Can use hclust instead:
#' spatial_clustering_cv(smithsonian_sf, v = 5, cluster_function = "hclust")
```

File: R/spatial_clustering_cv.R [hunk 10]
```diff
@@ -32,17 +38,42 @@
...
+    }
+    coords <- sf::st_centroid(sf::st_geometry(data))
+    coords <- as.dist(sf::st_distance(coords))
+  } else {
+    coords <- tidyselect::eval_select(rlang::enquo(coords), data = data)
+    if (is_empty(coords)) {
+      rlang::abort("`coords` are required and must be variables in `data`.")
+    }
+    coords <- data[coords]
+    coords <- dist(coords)
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:69]: This chunk isn't too long, but we could make a new generic function here to compute the `dist` (possible name such as `compute_coords_dist()`), with methods for a dataframe and `sf`. Do you think we would use something like this in another method as well? Or should we just stick with `if () {} else {}`?

File: R/spatial_clustering_cv.R [hunk 11]
```diff
@@ -32,17 +38,42 @@
...
   }
 
-  split_objs <- spatial_clustering_splits(data = data, coords = coords, v = v, ...)
+  split_objs <- spatial_clustering_splits(data = data,
+                                          coords = coords,
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:73]: Should we change this argument to `dist`? It seems kind of confusing now to call it `coords`.

File: R/spatial_clustering_cv.R [hunk 12]
```diff
@@ -59,16 +90,31 @@ spatial_clustering_cv <- function(data, coords, v = 10, ...) {
...
+
   if (!is.numeric(v) || length(v) != 1) {
     rlang::abort("`v` must be a single integer.")
   }
 
   n <- nrow(data)
-  clusters <- kmeans(data[coords], centers = v, ...)
-  folds <- clusters$cluster
+
+  folds <- switch(
```
Comment (juliasilge) [inline R/spatial_clustering_cv.R:106]: Great use of `switch()`!

**Review by juliasilge — CHANGES_REQUESTED** [review]
This is looking great! 🙌  I have just a few more small things. The most important is probably whether to make a new generic function for internal use in the resampling functions.

File: NEWS.md [hunk 13]
```diff
@@ -1,5 +1,19 @@
 # spatialsample (development version)
 
+* `spatial_clustering_cv()` gains an argument, `cluster_function`, which 
+  specifies what type of clustering to perform. `cluster_function = "kmeans"`, 
+  the default, uses [stats::kmeans()] for k-means clustering, while
```
Comment (mikemahoney218) [inline NEWS.md:5]:
```suggestion
  the default, uses `stats::kmeans()` for k-means clustering, while 
```

File: R/spatial_clustering_cv.R (hunk 12 above)
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R:107]: Thanks :smile:

File: R/spatial_clustering_cv.R (hunk 10 above)
Comment (mikemahoney218) [reply R/spatial_clustering_cv.R]: I don't _think_ we'll need this exact method again. I _believe_ the other methods we've talked about are going to rely pretty heavily on `sf` distance calculations, and less on the distance matrix for the entire data set, but I could be wrong there. Happy to extract it if we want though!

Comment (juliasilge) [reply R/spatial_clustering_cv.R]: Let's leave it as is for now. 👍

**Review by juliasilge — APPROVED** [review]
Hooray! 🚀

**Pull request merged**